pkg/volume/storageos/storageos_util.go

```diff
@@ -0,0 +1,433 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storageos
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/volume"
+
+	"github.com/golang/glog"
+	storageosapi "github.com/storageos/go-api"
+	storageostypes "github.com/storageos/go-api/types"
+)
+
+const (
+	losetupPath   = "losetup"
+	apiSecretName = "storageos-api"
+
+	modeBlock deviceType = iota
+	modeFile
+	modeUnsupported
+
+	ErrDeviceNotFound     = "device not found"
+	ErrDeviceNotSupported = "device not supported"
+	ErrNotAvailable       = "not available"
+)
+
+type deviceType int
+
+// storageosVolume describes a provisioned volume
+type storageosVolume struct {
+	ID          string
+	Name        string
+	Namespace   string
+	Description string
+	Pool        string
+	Size        int
+	Labels      map[string]string
+	FSType      string
+}
+
+type storageosAPIConfig struct {
+	apiAddr    string
+	apiUser    string
+	apiPass    string
+	apiVersion string
+}
+
+type apiConfigGetter interface {
+	GetAPIConfig() (*storageosAPIConfig, error)
+}
+
+type namespacedSecret struct {
+	namespace  string
+	secretName string
+	host       volume.VolumeHost
+}
+
+// GetAPIConfig checks the namespace for the api config.
+func (a *namespacedSecret) GetAPIConfig() (*storageosAPIConfig, error) {
+
+	c := &storageosAPIConfig{}
+
+	kubeClient := a.host.GetKubeClient()
+	if kubeClient == nil {
+		return nil, fmt.Errorf("Cannot get kube client")
+	}
+
+	keys, err := kubeClient.Core().Secrets(a.namespace).Get(a.secretName, metav1.GetOptions{})
```

Comment:
the pkg/volume/util#GetSecretForPV and GetSecretForPod helper functions enforce best practices (pulls namespace from Pod when fetching for a pod-referenced volume source, and ensures secret type matches when fetching for a PV-referenced volume source. Check how other plugins call those to see how they plumb the data from the source.